src/plugins/data/common/index_patterns/types.ts

```diff
@@ -51,3 +52,46 @@ export interface IndexPatternAttributes {
 
 export type OnNotification = (toastInputFields: ToastInputFields) => void;
 export type OnError = (error: Error, toastInputFields: ErrorToastOptions) => void;
+
+export interface FieldFormatSpec {
+  id: string;
+  params: IFieldFormatMetaParams;
+}
+
+export type AggregationRestrictions = Record<
+  string,
+  {
+    agg?: string;
+    interval?: number;
+    fixed_interval?: string;
+    calendar_interval?: string;
+    delay?: string;
+    time_zone?: string;
+  }
+>;
+
+export interface TypeMeta {
+  aggs?: Record<string, AggregationRestrictions>;
+  [key: string]: any;
+}
+
+export interface FieldSpec {
+  [key: string]: any;
+  format?: FieldFormatSpec;
```

Comment:
can we improve this type ? (list of all properties toSpec returns is known)

Reply:
I've taken a swing at this but its a much deeper refactor than it looks due to the poor state of type usage in all field related code. This PR has a well defined FieldSpec - https://github.com/elastic/kibana/pull/67817